ca/reconciler.go

```diff
@@ -0,0 +1,254 @@
+	wg                  *sync.WaitGroup
+
+	currentRootCA    *api.RootCA
+	currentIssuer    IssuerInfo
+	allNodes         map[string]*api.Node
+	unconvergedNodes map[string]struct{}
+	isReconciling    bool
+}
+
+func newReconciler(ctx context.Context, clusterID string, batchUpdateInteral time.Duration, s *store.MemoryStore, wg *sync.WaitGroup, rootCA *api.RootCA, nodes []*api.Node) *rootRotationReconciler {
+	r := &rootRotationReconciler{
+		ctx:                 ctx,
+		clusterID:           clusterID,
+		store:               s,
+		batchUpdateInterval: batchUpdateInteral,
+		wg:                  wg,
+		allNodes:            make(map[string]*api.Node),
+		unconvergedNodes:    make(map[string]struct{}),
+	}
+	r.UpdateRootCA(rootCA)
+	r.UpdateNodes(nodes...)
+	return r
+}
+
+func (r *rootRotationReconciler) UpdateRootCA(newRootCA *api.RootCA) {
+	if newRootCA == nil {
+		return
+	}
+	wantedIssuer := newRootCA.CACert
+	if newRootCA.RootRotation != nil {
+		wantedIssuer = newRootCA.RootRotation.CACert
+	}
+	issuerCerts, err := helpers.ParseCertificatesPEM(wantedIssuer)
+	if err != nil {
+		log.G(r.ctx).WithError(err).Error("invalid certificate in cluster root CA object")
+		return
+	}
+	if len(issuerCerts) == 0 {
+		log.G(r.ctx).Error("invalid certificate in cluster root CA object")
+		return
+	}
+
+	r.mu.Lock()
+	r.currentRootCA = newRootCA
+	if bytes.Equal(issuerCerts[0].RawSubject, r.currentIssuer.Subject) && bytes.Equal(issuerCerts[0].RawSubjectPublicKeyInfo, r.currentIssuer.PublicKey) {
+		r.mu.Unlock()
+		return
+	}
+	// If the issuer has changed, iterate through all the nodes to figure out which ones need rotation
+	r.currentIssuer = IssuerInfo{
+		Subject:   issuerCerts[0].RawSubject,
+		PublicKey: issuerCerts[0].RawSubjectPublicKeyInfo,
+	}
+	r.unconvergedNodes = make(map[string]struct{})
+	for _, n := range r.allNodes {
+		if hasIssuer(n, &r.currentIssuer) {
+			continue
+		}
+		r.unconvergedNodes[n.ID] = struct{}{}
+	}
+	hasRootRotation := r.currentRootCA.RootRotation != nil
+	r.mu.Unlock()
+
+	if hasRootRotation {
+		r.runReconcilerLoop()
+	}
+}
+
+func (r *rootRotationReconciler) UpdateNodes(nodes ...*api.Node) {
+	r.mu.Lock()
+	for _, n := range nodes {
+		if n == nil || n.Spec.Membership != api.NodeMembershipAccepted {
+			continue
+		}
+		r.allNodes[n.ID] = n
+		if hasIssuer(n, &r.currentIssuer) {
+			delete(r.unconvergedNodes, n.ID)
+		} else {
+			r.unconvergedNodes[n.ID] = struct{}{}
+		}
+	}
+	r.mu.Unlock()
+}
+
+func (r *rootRotationReconciler) DeleteNode(node *api.Node) {
+	if node == nil {
+		return
+	}
+	r.mu.Lock()
+	delete(r.allNodes, node.ID)
+	delete(r.unconvergedNodes, node.ID)
+	r.mu.Unlock()
+}
+
+func (r *rootRotationReconciler) runReconcilerLoop() {
+	r.mu.Lock()
+	if r.isReconciling { // loop is already going
+		r.mu.Unlock()
+		return
+	}
+	r.isReconciling = true
+	r.mu.Unlock()
+
+	r.wg.Add(1)
+	go func() {
+		defer func() {
+			r.wg.Done()
+			r.mu.Lock()
+			r.isReconciling = false
```

Comment:
Race: if `runReconcilerLoop` gets called after this goroutine decides to return, but before this lock is acquired and `isReconciling` is set to false.

One way to solve this would be to have a channel or context that can stop the goroutine, and to always stop an existing goroutine (if any) and then start a new one.